src/compiler/checker/lines_16000_18000.rs

```diff
@@ -132,6 +140,195 @@ impl TypeChecker {
+        if sources.len() == 1 {
+            self.make_unary_type_mapper(
+                sources[0].clone(),
+                if let Some(targets) = targets {
+                    targets[0].clone()
+                } else {
+                    self.any_type()
+                },
+            )
+        } else {
+            self.make_array_type_mapper(sources, targets)
+        }
+    }
+
+    pub(super) fn get_mapped_type(&self, type_: Rc<Type>, mapper: &TypeMapper) -> Rc<Type> {
+        match mapper {
+            TypeMapper::Simple(mapper) => {
+                if Rc::ptr_eq(&type_, &mapper.source) {
+                    mapper.target.clone()
+                } else {
+                    type_
+                }
+            }
+            TypeMapper::Array(mapper) => {
+                let sources = &mapper.sources;
+                let targets = &mapper.targets;
+                for (i, source) in sources.iter().enumerate() {
+                    if Rc::ptr_eq(&type_, source) {
+                        return if let Some(targets) = targets {
+                            targets[i].clone()
+                        } else {
+                            self.any_type()
+                        };
+                    }
+                }
+                type_
+            }
+            TypeMapper::Function(mapper) => (mapper.func)(self, type_),
+            TypeMapper::Composite(composite_or_merged_mapper)
+            | TypeMapper::Merged(composite_or_merged_mapper) => {
+                let t1 = self.get_mapped_type(type_.clone(), &composite_or_merged_mapper.mapper1);
+                if !Rc::ptr_eq(&t1, &type_) && matches!(mapper, TypeMapper::Composite(_)) {
+                    self.instantiate_type(Some(t1), Some(&composite_or_merged_mapper.mapper2))
+                        .unwrap()
+                } else {
+                    self.get_mapped_type(t1, &composite_or_merged_mapper.mapper2)
+                }
+            }
+        }
+    }
+
+    pub(super) fn make_unary_type_mapper(&self, source: Rc<Type>, target: Rc<Type>) -> TypeMapper {
+        TypeMapper::new_simple(source, target)
+    }
+
+    pub(super) fn make_array_type_mapper(
+        &self,
+        sources: Vec<Rc<Type /*TypeParameter*/>>,
+        targets: Option<Vec<Rc<Type>>>,
+    ) -> TypeMapper {
+        TypeMapper::new_array(sources, targets)
+    }
+
+    pub(super) fn make_composite_type_mapper(
+        &self,
+        mapper1: TypeMapper,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        TypeMapper::new_composite(mapper1, mapper2)
+    }
+
+    pub(super) fn make_merged_type_mapper(
+        &self,
+        mapper1: TypeMapper,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        TypeMapper::new_merged(mapper1, mapper2)
+    }
+
+    pub(super) fn combine_type_mappers(
+        &self,
+        mapper1: Option<TypeMapper>,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        if let Some(mapper1) = mapper1 {
```

Comment:
Same comment as above about `map_or_unwrap`.

Reply:
In this case it's unhappy that I'm trying to "own" `mapper2` twice if I use `.map_or()`/`.map_or_else()` since I guess it's not smart enough to know that eg in `.map_or_else()` only one of the two closures will be executed